File: BaekJoon/13975_파일합치기3_PriorityQueue.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;
using ll = long long;

// 파일합치기3 13975 PriorityQueue
// O( N logN )


int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int t;
    cin >> t;

    while (t--)
    {
        int k;
        cin >> k;

        // 최소힙
        priority_queue<ll, vector<ll>, greater<ll>> pq;

        for (int i = 0; i < k; ++i)
        {
            ll f;
            cin >> f;
            pq.push(f);
        }

        ll totalCost{ 0 };  // 병합에 든 총 비용
        while (pq.size() > 1)
        {
            ll first = pq.top(); pq.pop();
            ll second = pq.top(); pq.pop();

            // 가장 작은 두 값을 병합
            ll merge = first + second;
            totalCost += merge; // 총 비용에 +
            pq.push(merge); // 병합한 값을 다시 push
        }

        cout << totalCost << "\n";
    }


    return 0;
}

File: BaekJoon/Java/13975_파일합치기3_PriorityQueue.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

// 파일합치기3 13975 PriorityQueue
// O( N logN )


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = null;

        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            int k = Integer.parseInt(br.readLine());

            PriorityQueue<Long> pq = new PriorityQueue<>(k);

            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < k; ++i) {
                pq.add(Long.parseLong(st.nextToken()));
            }

            long totalCost = 0;
            while (pq.size() > 1) {
                long first = pq.poll();
                long second = pq.poll();

                long merge = first + second;
                totalCost += merge;
                pq.add(merge);
            }

            System.out.println(totalCost);
        }
    }
}
